apps/testing-javascript/src/data/purchase-import.ts

```diff
@@ -0,0 +1,260 @@
+    usedCount: z.number(),
+    percentageDiscount: z.string(),
+    restrictedToProductId: z.string().nullable(),
+    bulkPurchaseId: z.string().nullable(),
+  })
+  .transform((coupon) => {
+    return {
+      ...coupon,
+      percentageDiscount: Number.parseFloat(coupon.percentageDiscount),
+      maxUses: coupon.maxUses || undefined,
+    }
+  })
+  .transform(({createdAt, expires, ...rest}) => {
+    return {
+      createdAt: new Date(createdAt),
+      expires: expires && new Date(expires),
+      ...rest,
+    }
+  })
+
+const MerchantCustomerSchema = z
+  .object({
+    id: z.string(),
+    userId: z.string(),
+    merchantAccountId: z.string(),
+    identifier: z.string(),
+    createdAt: z.string(),
+    status: z.number(),
+  })
+  .transform(({createdAt, ...rest}) => {
+    return {
+      createdAt: new Date(createdAt),
+      ...rest,
+    }
+  })
+
+const MerchantChargeSchema = z
+  .object({
+    id: z.string(),
+    status: z.number(),
+    identifier: z.string(),
+    userId: z.string(),
+    merchantAccountId: z.string(),
+    merchantProductId: z.string(),
+    merchantCustomerId: z.string().nullable(),
+  })
+  .transform((merchantCharge) => {
+    return {
+      ...merchantCharge,
+      merchantCustomerId: merchantCharge.merchantCustomerId || undefined,
+    }
+  })
+
+type MerchantCharge = z.infer<typeof MerchantChargeSchema>
+
+// from: https://github.com/sindresorhus/type-fest/blob/main/source/set-non-nullable.d.ts
+type SetNonNullable<BaseType, Keys extends keyof BaseType = keyof BaseType> = {
+  [Key in keyof BaseType]: Key extends Keys
+    ? NonNullable<BaseType[Key]>
+    : BaseType[Key]
+}
+
+type CleanMerchantCharge = SetNonNullable<MerchantCharge, 'merchantCustomerId'>
+
+const PurchaseSchema = z
+  .object({
+    id: z.string(),
+    userId: z.string(),
+    createdAt: z.string(),
+    totalAmount: z.coerce.number(),
+    city: z.string().nullable(),
+    state: z.string().nullable(),
+    country: z.string().nullable(),
+    ip_address: z.string().nullable(),
+    status: z.union([
+      z.literal('Valid'),
+      z.literal('Refunded'),
+      z.literal('Banned'),
+    ]),
+    productId: z.string(),
+    couponId: z.string().nullable(),
+    merchantChargeId: z.string().nullable(),
+    upgradedFromId: z.string().nullable(),
+    bulkCouponId: z.string().nullable(),
+    redeemedBulkCouponId: z.string().nullable(),
+    merchantPurchaseId: z.string().nullable(),
+    // quantity: z.number().nullable(),
+  })
+  .transform(({createdAt, merchantPurchaseId, ...rest}) => {
+    return {
+      createdAt: new Date(createdAt),
+      merchantSessionId: merchantPurchaseId,
+      ...rest,
+    }
+  })
+
+const FileDataSchema = z.object({
+  users: z.array(UserSchema),
+  coupons: z.array(CouponSchema),
+  merchantCustomers: z.array(MerchantCustomerSchema),
+  merchantCharges: z.array(MerchantChargeSchema),
+  purchases: z.array(PurchaseSchema),
+})
+
+const importPurchaseData = async () => {
```

Comment:
This script runs through a JSON file of exported `User`, `Coupon`, `MerchantCustomer`, `MerchantCharge`, and `Purchase` data and does batched creates against the MySQL database at `DATABASE_URL`.